### ISSUE TITLE
💅 noLeakedRender false positive when using useState<boolean>

### ISSUE THREAD

**Issue by plaa** [biomejs/biome#10790]
### Environment information

<details>

```bash

CLI:
  Version:                      2.5.0
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           linux

Environment:
  BIOME_DISTRIBUTION:                npm
  BIOME_LOG_PATH:                    unset
  BIOME_LOG_PREFIX_NAME:             unset
  BIOME_LOG_LEVEL:                   unset
  BIOME_LOG_KIND:                    unset
  BIOME_CONFIG_PATH:                 unset
  BIOME_THREADS:                     unset
  BIOME_WATCHER_KIND:                unset
  BIOME_WATCHER_POLLING_INTERVAL:    unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v24.15.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         npm/11.12.1

Biome Configuration:
  Status:                       Loaded successfully.
  Path:                         /workspaces/ecommerce/biome.jsonc
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false
  HTML full support enabled:    unset

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  unset
  Enabled rules:
    a11y/noAccessKey
    a11y/noAmbiguousAnchorText
    a11y/noAriaHiddenOnFocusable
    a11y/noAriaUnsupportedElements
    a11y/noAutofocus
    a11y/noDistractingElements
    a11y/noHeaderScope
    a11y/noInteractiveElementToNoninteractiveRole
    a11y/noLabelWithoutControl
    a11y/noNoninteractiveElementInteractions
    a11y/noNoninteractiveElementToInteractiveRole
    a11y/noNoninteractiveTabindex
    a11y/noPositiveTabindex
    a11y/noRedundantAlt
    a11y/noRedundantRoles
    a11y/noStaticElementInteractions
    a11y/noSvgWithoutTitle
    a11y/useAltText
    a11y/useAnchorContent
    a11y/useAriaActivedescendantWithTabindex
    a11y/useAriaPropsForRole
    a11y/useAriaPropsSupportedByRole
    a11y/useButtonType
    a11y/useFocusableInteractive
    a11y/useGenericFontNames
    a11y/useHeadingContent
    a11y/useHtmlLang
    a11y/useIframeTitle
    a11y/useKeyWithClickEvents
    a11y/useKeyWithMouseEvents
    a11y/useMediaCaption
    a11y/useSemanticElements
    a11y/useValidAnchor
    a11y/useValidAriaProps
    a11y/useValidAriaRole
    a11y/useValidAriaValues
    a11y/useValidAutocomplete
    a11y/useValidLang
    complexity/noAdjacentSpacesInRegex
    complexity/noArguments
    complexity/noBannedTypes
    complexity/noCommaOperator
    complexity/noEmptyTypeParameters
    complexity/noExcessiveNestedTestSuites
    complexity/noExtraBooleanCast
    complexity/noFlatMapIdentity
    complexity/noImportantStyles
    complexity/noStaticOnlyClass
    complexity/noThisInStatic
    complexity/noUselessCatch
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessStringConcat
    complexity/noUselessStringRaw
    complexity/noUselessSwitchCase
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noVoid
    complexity/useArrowFunction
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useIndexOf
    complexity/useLiteralKeys
    complexity/useNumericLiterals
    complexity/useOptionalChain
    complexity/useRegexLiterals
    complexity/useSimpleNumberKeys
    complexity/useWhile
    correctness/noBeforeInteractiveScriptOutsideDocument
    correctness/noChildrenProp
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noDuplicateArgumentNames
    correctness/noDuplicateAttributes
    correctness/noDuplicateEnumValueNames
    correctness/noDuplicateInputFieldNames
    correctness/noDuplicateVariableNames
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalDirnameFilename
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNestedComponentDefinitions
    correctness/noNextAsyncClientComponent
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noPrivateImports
    correctness/noReactPropAssignments
    correctness/noRenderReturnValue
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noStringCaseMismatch
    correctness/noSwitchDeclarations
    correctness/noUndeclaredDependencies
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnresolvedImports
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/useExhaustiveDependencies
    correctness/useGraphqlNamedOperations
    correctness/useHookAtTopLevel
    correctness/useImportExtensions
    correctness/useInlineScriptId
    correctness/useIsNan
    correctness/useJsonImportAttributes
    correctness/useJsxKeyInIterable
    correctness/useLoneAnonymousOperation
    correctness/useParseIntRadix
    correctness/useUniqueElementIds
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    nursery/noFloatingPromises
    nursery/noMisusedPromises
    nursery/useExhaustiveSwitchCases
    performance/noAccumulatingSpread
    performance/noAwaitInLoops
    performance/noBarrelFile
    performance/noDynamicNamespaceImportAccess
    performance/noImgElement
    performance/noJsxPropsBind
    performance/noReExportAll
    performance/noSyncScripts
    performance/noUnwantedPolyfillio
    performance/useGoogleFontPreconnect
    security/noBlankTarget
    security/noDangerouslySetInnerHtml
    security/noDangerouslySetInnerHtmlWithChildren
    security/noGlobalEval
    security/noScriptUrl
    style/noDescendingSpecificity
    style/noHeadElement
    style/noInferrableTypes
    style/noNamespace
    style/noNonNullAssertion
    style/noParameterAssign
    style/noParameterProperties
    style/noRestrictedImports
    style/noUnusedTemplateLiteral
    style/noYodaExpression
    style/useArrayLiterals
    style/useAsConstAssertion
    style/useBlockStatements
    style/useComponentExportOnlyModules
    style/useConsistentCurlyBraces
    style/useConsistentObjectDefinitions
    style/useConst
    style/useDefaultParameterLast
    style/useDefaultSwitchClause
    style/useDeprecatedReason
    style/useEnumInitializers
    style/useExplicitLengthCheck
    style/useExponentiationOperator
    style/useExportType
    style/useForOf
    style/useGroupedAccessorPairs
    style/useLiteralEnumMembers
    style/useNamingConvention
    style/useNodejsImportProtocol
    style/useNumberNamespace
    style/useObjectSpread
    style/useReactFunctionComponents
    style/useSelfClosingElements
    style/useShorthandFunctionType
    style/useTemplate
    style/useThrowNewError
    style/useThrowOnlyError
    style/useUnifiedTypeSignatures
    suspicious/noApproximativeNumericConstant
    suspicious/noArrayIndexKey
    suspicious/noAssignInExpressions
    suspicious/noAsyncPromiseExecutor
    suspicious/noBiomeFirstException
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConfusingVoidType
    suspicious/noConstEnum
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDeprecatedImports
    suspicious/noDeprecatedMediaType
    suspicious/noDocumentCookie
    suspicious/noDocumentImportInPage
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateEnumValues
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noDuplicateTestHooks
    suspicious/noDuplicatedSpreadProps
    suspicious/noEmptyBlock
    suspicious/noEmptyInterface
    suspicious/noExplicitAny
    suspicious/noExportsInTest
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFocusedTests
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noGlobalIsNan
    suspicious/noHeadImportInDocument
    suspicious/noImplicitAnyLet
    suspicious/noImportAssign
    suspicious/noImportCycles
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noLeakedRender
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noNonNullAssertedOptionalChain
    suspicious/noOctalEscape
    suspicious/noProto
    suspicious/noPrototypeBuiltins
    suspicious/noQuickfixBiome
    suspicious/noReactForwardRef
    suspicious/noRedeclare
    suspicious/noRedundantUseStrict
    suspicious/noSelfCompare
    suspicious/noShadowRestrictedNames
    suspicious/noShorthandPropertyOverrides
    suspicious/noSkippedTests
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noTemplateCurlyInString
    suspicious/noThenProperty
    suspicious/noTsIgnore
    suspicious/noUnknownAtRules
    suspicious/noUnknownAttribute
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noUselessEscapeInString
    suspicious/noUselessRegexBackrefs
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useBiomeIgnoreFolder
    suspicious/useDefaultSwitchClauseLast
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useIsArray
    suspicious/useIterableCallbackReturn
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0
```

</details>

### Rule name

lint/suspicious/noLeakedRender

### Playground link

https://biomejs.dev/playground/?lineWidth=120&indentStyle=space&lintRules=noLeakedRender&tab=syntax&pane=Diagnostics&code=ZQB4AHAAbwByAHQAIABkAGUAZgBhAHUAbAB0ACAAZgB1AG4AYwB0AGkAbwBuACAATQB5AEMAbwBtAHAAbwBuAGUAbgB0ACgAKQAgAHsACgAgACAAYwBvAG4AcwB0ACAAWwBlAHIAcgBvAHIALAAgAHMAZQB0AEUAcgByAG8AcgBdACAAPQAgAHUAcwBlAFMAdABhAHQAZQAoAGYAYQBsAHMAZQApADsACgAgACAALwAvACAAVABoAGkAcwAgAGQAbwBlAHMAIABuAG8AdAAgAGMAYQB1AHMAZQAgAHQAaABlACAAdwBhAHIAbgBpAG4AZwA6AAoAIAAgAC8ALwBjAG8AbgBzAHQAIABlAHIAcgBvAHIAIAA9ACAAZgBhAGwAcwBlADsACgAKACAAIAByAGUAdAB1AHIAbgAgADwAPgAKACAAIAAgACAAewBlAHIAcgBvAHIAIAAmACYAIAAiAEUAcgByAG8AcgAgAG8AYwBjAHUAcgByAGUAZAAiAH0ACgAgACAAPAAvAD4ACgB9AA%3D%3D

### Expected result

`useState` boolean values are often used as conditions in rendering, they should not cause the lint error.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by dyc3** [comment]
I'm not convinced this is a false positive, and not just a limitation of the rule. The source rule doesn't use type information and I'm pretty sure it would flag this case: https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-no-leaked-render.md

**Comment by plaa** [comment]
The Biome rule does use type information since using `const error = false` does not produce the warning, i.e. it knows that `error` is of `boolean` type.  However when using `useState` (which is a very common pattern for such conditional rendering) it apparently doesn't correctly infer the type.

We would value having this rule enabled, but it caused 60 false positives (through usage of `useState`) in our code base, so we ended up disabling the rule.

The inspiration `react/jsx-no-leaked-render` rule does state at the bottom of the page:

> If you are working in a typed-codebase which encourages you to always use boolean conditions, this rule can be disabled.

However, "encouraging to use boolean conditions" isn't the same as linting that it is actually the case.  All the invalid cases listed on [the Biome rule's page](https://biomejs.dev/linter/rules/no-leaked-render/) could still occur in a typed codebase.  Biome can also strive to be better than eslint.

I fully understand if this kind of type inference would be technically too difficult to achieve, in which case this can be closed as a "won't fix" bug, but I hope it's acknowledged that these are false positives and that it would be preferable to fix this.

**Comment by dyc3** [comment]
> The Biome rule does use type information

To be clear, our rule doesn't use type information right now. [It only uses our semantic model.](https://github.com/biomejs/biome/blob/3f9e3c959a900dc6c0cbd3a9aebc59274c2dad5d/crates/biome_js_analyze/src/lint/suspicious/no_leaked_render.rs#L103)

It is a difficult problem. I would personally defer to the source rule's advice: disable the rule.

But I'm not a react guy, so I don't really have the best context on this rule.
